resolver-dns/src/main/java/io/netty/resolver/dns/UnixResolverDnsServerAddressStreamProvider.java

```diff
@@ -0,0 +1,233 @@
+            throw new IllegalArgumentException("no files to parse");
+        }
+        if (etcResolverFiles != null) {
+            domainToNameServerStreamMap = parse(etcResolverFiles);
+            if (etcResolvConf != null) {
+                Map<String, DnsServerAddresses> etcResolvConfMap = parse(etcResolvConf);
+                defaultNameServerAddresses = etcResolvConfMap.remove(etcResolvConf.getName());
+                domainToNameServerStreamMap.putAll(etcResolvConfMap);
+            } else {
+                defaultNameServerAddresses = null;
+            }
+        } else {
+            domainToNameServerStreamMap = parse(etcResolvConf);
+            defaultNameServerAddresses = domainToNameServerStreamMap.remove(etcResolvConf.getName());
+        }
+    }
+
+    /**
+     * Parse a file of the format <a href="https://linux.die.net/man/5/resolver">/etc/resolv.conf</a> which may contain
+     * the default DNS server to use, and also overrides for individual domains. Also parse a directory of the format
+     * <a href="
+     * https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man5/resolver.5.html">
+     * /etc/resolver</a> which may contain multiple files to override the name servers used for multimple domains.
+     * @param etcResolvConf <a href="https://linux.die.net/man/5/resolver">/etc/resolv.conf</a>.
+     * @param etcResolverDir Directory containing files of the format defined in
+     * <a href="
+     * https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man5/resolver.5.html">
+     * /etc/resolver</a>.
+     * @throws IOException If an error occurs while parsing the input files.
+     */
+    public UnixResolverDnsServerAddressStreamProvider(String etcResolvConf, String etcResolverDir) throws IOException {
+        this(etcResolvConf == null ? null : new File(etcResolvConf),
+             etcResolverDir == null ? null :new File(etcResolverDir).listFiles());
+    }
+
+    @Override
+    public DnsServerAddressStream nameServerAddressStream(String hostname) {
+        for (;;) {
+            int i = hostname.indexOf('.', 1);
+            if (i < 0 || i == hostname.length() - 1) {
+                return defaultNameServerAddresses != null ? defaultNameServerAddresses.stream() : null;
+            }
+
+            DnsServerAddresses addresses = domainToNameServerStreamMap.get(hostname);
+            if (addresses != null) {
+                return addresses.stream();
+            }
+
+            hostname = hostname.substring(i + 1);
+        }
+    }
+
+    boolean mayOverrideNameServers() {
+        return !domainToNameServerStreamMap.isEmpty() ||
+                defaultNameServerAddresses != null && defaultNameServerAddresses.stream().next() != null;
+    }
+
+    private static Map<String, DnsServerAddresses> parse(File... etcResolverFiles) throws IOException {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "parse" is 26 which is greater than 25 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)